File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/aas/backend/connected/aas/submodelelement/property/blob/ConnectedBlob.java
package org.eclipse.basyx.aas.backend.connected.aas.submodelelement.property.blob;

import java.util.List;

import org.eclipse.basyx.aas.api.metamodel.aas.submodelelement.property.blob.IBlob;
import org.eclipse.basyx.aas.backend.connected.aas.submodelelement.ConnectedDataElement;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.submodelelement.property.Property;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.submodelelement.property.blob.Blob;
import org.eclipse.basyx.vab.core.proxy.VABElementProxy;
/**
 * "Connected" implementation of IBlob
 * @author rajashek
 *
 */
public class ConnectedBlob extends ConnectedDataElement implements IBlob {
	public ConnectedBlob(VABElementProxy proxy) {
		super(proxy);		
	}
	
	@Override
	public void setValue(byte[] value) {
		getProxy().setModelPropertyValue(Property.VALUE, value);
		
	}

	@SuppressWarnings("unchecked")
	@Override
	public byte[] getValue() {
		return convert((List<Integer>) getProxy().getModelPropertyValue(Property.VALUE));
	}

	@Override
	public void setMimeType(String mimeType) {
		getProxy().setModelPropertyValue(Blob.MIMETYPE, mimeType);
		
	}

	@Override
	public String getMimeType() {
		return (String) getProxy().getModelPropertyValue(Blob.MIMETYPE);
	}

	/**
	 * Converts a list of integers to a byte array. <br />
	 * Needed since the VAB does only support lists
	 * 
	 * @param array
	 * @return
	 */
	private byte[] convert(List<Integer> array) {
		byte[] res = new byte[array.size()];
		for (int i = 0; i < array.size(); i++) {
			res[i] = array.get(i).byteValue();
		}
		return res;
	}
}

File: sdks/c++/basys.sdk.cc/src/aas/backend/connected/aas/submodelelement/property/blob/ConnectedBlob.cpp
/*
 * ConnectedBlob.cpp
 *
 *      Author: wendel
 */

#include "ConnectedBlob.h"
#include "aas/submodelelement/property/IProperty.h"

namespace basyx {
namespace aas {
namespace backend {
namespace connected {

ConnectedBlob::ConnectedBlob(std::shared_ptr<vab::core::proxy::IVABElementProxy> proxy) :
 ConnectedDataElement(proxy)
{}

void ConnectedBlob::setValue(const std::string & bytes)
{
  this->setProxyValue(submodelelement::property::PropertyPaths::VALUE, bytes);
}

std::string ConnectedBlob::getValue() const
{
  auto value = getProxy()->readElementValue(submodelelement::property::PropertyPaths::VALUE);
  return value.Get<std::string>();
}

void ConnectedBlob::setMimeType(const std::string & mimeType)
{
  this->setProxyValue(submodelelement::BlobPath::MIMETYPE, mimeType);
}

std::string ConnectedBlob::getMimeType() const
{
  return this->getProxyValue(submodelelement::BlobPath::MIMETYPE);
}

}
}
}
}
